### PR TITLE
Add BPM tags to built-in beat loops (#5439)

### PR DESCRIPTION
**Pull request by mirk0dex**
- Added floating-point vorbis BPM tags to files in lmms/data/samples/beats
- Added rounded BPM to filenames, surrounded by square brackets and separated from the rest of the filename by an underscore, as requested by @ttepatti.

To achieve this, I used the [bpm-tools](https://www.pogo.org.uk/~mark/bpm-tools) utils and verified that the values were correct using LMMS itself. When the tempos were not correct, I fired up the DAW and started looking, manually, for one fitting the loop. When I found one, I once again used bpm-tools (but this time with a range that suited the approximate tempo) and detected the precise value.

I then proceed to add the vorbis tags, and rename the files. 

All tempos should be accurate. Only loops I'd double-check are the breaks: finding the right BPM wasn't easy there.

### CODE REVIEW

**Comment by mirk0dex** [comment]
I found out that LMMS has an "assets" repo. Gonna use that.

**Comment by mirk0dex** [comment]
Not sure whether it's OK for me to leave this here or not. I can't find the modified loops in <https://github.com/LMMS/assets>, so I'm keeping the PR open.

**Comment by zonkmachine** [comment]
Hi, thanks for looking into this!

Now that the names are changed we need some form of upgrade routine so old projects that use the samples will find them. How is your coding? :)

**Comment by mirk0dex** [comment]
Hey, I'd say it's fine.

I'd be glad to help out, if you tell me what to do I'll try my best to wrap my head around it.

**Comment by zonkmachine** [comment]
OK. When we need to upgrade a project file to work with a new feature, name changes quite commonly, the job is done here: https://github.com/LMMS/lmms/blob/master/src/core/DataFile.cpp

Here is a recent PR that addressed an issue with elements that had been renamed but the upgrade didn't work:
https://github.com/LMMS/lmms/pull/6851

Here is a minimal demo project with just a loop that needs to be upgraded to the new name
[beatupgrade.mmp.txt](https://github.com/LMMS/lmms/files/12641568/beatupgrade.mmp.txt)
`<sampleclip muted="0" off="0" src="factorysample:beats/electro_beat01.ogg" pos="0" sample_rate="44100" len="384"/>`

Have fun!

**Comment by mirk0dex** [comment]
Thanks! I'll look into this tomorrow.

**Comment by mirk0dex** [comment]
I created an upgrade procedure. It simply loops through all <sampleclip> elements and if they contain one of the changed strings, it appends " - X BPM" to the sample names.

Unfortunately, it looks like Emacs drastically changed the file's indentation and code arrangement, for some reason. If you want me to, I can restore the original version and use vim or something to copy and paste my function in it.

Also, I haven't tested any of the code yet, because I'm currently using Void GNU/Linux, and building LMMS here is a pain in the neck.

**Comment by zonkmachine** [comment]
> If you want me to, I can restore the original version and use vim or something to copy and paste my function in it.

Want. The problem is that it's one commit and it's not easy to find your intended changes among all the format changes. The format changes, at a glance, do look like they follow the coding standard.

**Comment by mirk0dex** [comment]
> If you want me to, I can restore the original version and use vim or something to copy and paste my function in it.

Done.

I pasted it at the end of the file. I also committed it to the correct branch this time.

**Comment by zonkmachine** [comment]
> > If you want me to, I can restore the original version and use vim or something to copy and paste my function in it.
> 
> Done.
> 
> I pasted it at the end of the file. I also committed it to the correct branch this time.

Did this actually compile? You can't just add a function to DataFile.cpp without declaring it in DataFile.h
Search **src/core/Datafile.cpp** and **include/DataFile.h** for occurences of **upgrade_sampleAndHold()**. Try and *mirror what's done and stick your stuff after.

File: src/core/DataFile.cpp [hunk 1]
```diff
@@ -1960,5 +1960,64 @@ unsigned int DataFile::legacyFileVersion()
 	return std::distance( UPGRADE_VERSIONS.begin(), firstRequiredUpgrade );
 }
 
+// Change loops' filenames in <sampleclip>s
+void DataFile::upgrade_loopsRename()
+{
```
Comment (zonkmachine) [inline src/core/DataFile.cpp:1965]: Now this needs to be moved to after SampleAndHold routine. About here: https://github.com/LMMS/lmms/blob/de332820f7ed5ca33d64ffd62b8da147f6105c9a/src/core/DataFile.cpp#L1812

**Comment by zonkmachine** [comment]
@mirk0dex Does this actually compile for you?

```make
/builds/lmms/src/core/DataFile.cpp:1966:39: error: template argument 1 is invalid
 1966 |         std::vector<std::pair<string, string>> loopBPMs{
```

**Comment by mirk0dex** [comment]
> @mirk0dex Does this actually compile for you?
> 
> ```makefile
> /builds/lmms/src/core/DataFile.cpp:1966:39: error: template argument 1 is invalid
>  1966 |         std::vector<std::pair<string, string>> loopBPMs{
> ```

I haven't tried compiling anything, yet. Is there any way I can just compile that file, or check for errors without having to build everything?

**Comment by mirk0dex** [comment]
Maybe using std::string instead of just string will do the trick?

**Comment by zonkmachine** [comment]
> > @mirk0dex Does this actually compile for you?
> > ```makefile
> > /builds/lmms/src/core/DataFile.cpp:1966:39: error: template argument 1 is invalid
> >  1966 |         std::vector<std::pair<string, string>> loopBPMs{
> > ```
> 
> I haven't tried compiling anything, yet. Is there any way I can just compile that file, or check for errors without having to build everything?

No. You should make sure that things build before committing it to a project or be sure to mention that you haven't tested it yet.

**Comment by mirk0dex** [comment]
> > > @mirk0dex Does this actually compile for you?
> > > ```makefile
> > > /builds/lmms/src/core/DataFile.cpp:1966:39: error: template argument 1 is invalid
> > >  1966 |         std::vector<std::pair<string, string>> loopBPMs{
> > > ```
> > 
> > 
> > I haven't tried compiling anything, yet. Is there any way I can just compile that file, or check for errors without having to build everything?
> 
> No. You should make sure that things build before committing it to a project or be sure to mention that you haven't tested it yet.

I did say it in a previous message. Apologies for the inconvenience. Anyways, tomorrow I will have my Arch system with me, so I will be able to compile everything with ease.

**Comment by zonkmachine** [comment]
> Maybe using std::string instead of just string will do the trick?

QString is what the rest of the file is using. Stick to that. And then compile it. You actually need to do that. If you have problem compiling it you need to bring it up at our Discord site. This is not the right place.

**Comment by mirk0dex** [comment]
> > Maybe using std::string instead of just string will do the trick?
> 
> QString is what the rest of the file is using. Stick to that. And then compile it. You actually need to do that. If you have problem compiling it you need to bring it up at our Discord site. This is not the right place.

OK, will do.

**Comment by mirk0dex** [comment]
I tested the function. It now works as it should.

<sampleclip off="0" sample_rate="44100" len="384" muted="0" src="factorysample:beats/**electro_beat01 - 120 BPM.ogg**" pos="0"/>

**Comment by zonkmachine** [comment]
> I tested the function. It now works as it should.
> 
> `<sampleclip off="0" sample_rate="44100" len="384" muted="0" src="factorysample:beats/electro_beat01 - 120 BPM.ogg" pos="0"/>`

Right. This now upgrades. Loops can be imported into our builtin sampler, the audio file processor. This doesn't seem to upgrade.

```xml
          <instrument name="audiofileprocessor">
            <audiofileprocessor eframe="1" lframe="0" reversed="0" amp="100" stutter="0" interp="1" src="factorysample:beats/909beat01.ogg" looped="0" sframe="0">
              <key/>
            </audiofileprocessor>
          </instrument>
```

New project to test. I forgot to mention that we add .txt to the end of project files instead of zipping as github doesn't recognize .mmp. Just drop the suffix and go...
[beatupgrade2.mmp.txt](https://github.com/LMMS/lmms/files/12664115/beatupgrade2.mmp.txt)

File: src/core/DataFile.cpp [hunk 2]
```diff
@@ -1804,7 +1804,80 @@ void DataFile::upgrade_sampleAndHold()
...
+					{
+						auto currentPair = LOOPBPMS.at(k);
+						auto toReplace = currentPair.first;
+						auto srcVal = item.attribute(srcAttr);
+
+						if (srcVal.contains(toReplace, Qt::CaseInsensitive))
+						{
+							// Add " - X BPM" to filename
+							srcVal.replace(toReplace,
+									toReplace + BPMPREF + currentPair.second + BPMSUFF);
```
Comment (zonkmachine) [inline src/core/DataFile.cpp:1874]: The lines above can be written together. We go to 120 characters for a line.

File: include/DataFile.h [hunk 3]
```diff
@@ -125,6 +125,7 @@ class LMMS_EXPORT DataFile : public QDomDocument
 	void upgrade_extendedNoteRange();
 	void upgrade_defaultTripleOscillatorHQ();
 	void upgrade_mixerRename();
+	void upgrade_loopsRename();
```
Comment (zonkmachine) [inline include/DataFile.h:128]: Please move this two lines down to after `void upgrade_sampleAndHold();`

**Review by zonkmachine — APPROVED** [review]
I left some remarks on coding style but this looks good and tests out fine. The Vorbis tags look fine although I don't know the usefulness of bpm measured to 3 decimals precision. The upgrade works. I've run out of things to test now. Approved.

File: src/core/DataFile.cpp [hunk 4]
```diff
@@ -1804,7 +1804,80 @@ void DataFile::upgrade_sampleAndHold()
...
+			auto srcAttr = srcAttrs.at(i);
+
+			for (int j = 0; j < elements.length(); ++j)
+			{
+				auto item = elements.item(j).toElement();
+
+				if (item.isNull()) { continue; }
+				if (item.hasAttribute(srcAttr))
+				{
+					for (int k = 0; k < LOOPBPMS.size(); ++k)
```
Comment (DomClark) [inline src/core/DataFile.cpp:1864]: The `i` and `k` for-loops would be better as range-based for-loops. This `k` one could use a structured binding like the outermost for-loop.

File: src/core/DataFile.cpp [hunk 5]
```diff
@@ -1804,7 +1804,80 @@ void DataFile::upgrade_sampleAndHold()
...
+		{"latin_brass01", "140"},
+		{"latin_guitar01", "126"},
+		{"latin_guitar02", "140"},
+		{"latin_guitar03", "120"},
+	};
+
+	const QString BPMPREF = " - ", BPMSUFF = " BPM";
+
+	// Replace loop sample names
+	for (auto& [elem, srcAttrs] : ELEMENTS_WITH_RESOURCES)
```
Comment (DomClark) [inline src/core/DataFile.cpp:1849]: While `const` is automatically deduced in the types of `elem` and `srcAttrs`, I would rather it were explicit:
```suggestion
	for (const auto& [elem, srcAttrs] : ELEMENTS_WITH_RESOURCES)
```

File: src/core/DataFile.cpp [hunk 6]
```diff
@@ -1804,7 +1804,80 @@ void DataFile::upgrade_sampleAndHold()
...
+		{"rave_kick01", "180"},
+		{"rave_kick02", "180"},
+		{"rave_snare01", "180"},
+		{"latin_brass01", "140"},
+		{"latin_guitar01", "126"},
+		{"latin_guitar02", "140"},
+		{"latin_guitar03", "120"},
+	};
+
+	const QString BPMPREF = " - ", BPMSUFF = " BPM";
```
Comment (DomClark) [inline src/core/DataFile.cpp:1846]: I don't think there's any advantage to storing these strings in variables, seeing as they're used exactly once.

File: src/core/DataFile.cpp [hunk 7]
```diff
@@ -1804,7 +1804,80 @@ void DataFile::upgrade_sampleAndHold()
...
+				if (item.isNull()) { continue; }
+				if (item.hasAttribute(srcAttr))
+				{
+					for (int k = 0; k < LOOPBPMS.size(); ++k)
+					{
+						auto currentPair = LOOPBPMS.at(k);
+						auto toReplace = currentPair.first;
+						auto srcVal = item.attribute(srcAttr);
+
+						if (srcVal.contains(toReplace, Qt::CaseInsensitive))
```
Comment (DomClark) [inline src/core/DataFile.cpp:1870]: I worry that this check is too loose - we could easily end up replacing names of non-factory samples too. Can it be restricted further, to only match entire file names, and only in the factory sample directory?

**Comment by mirk0dex** [comment]
Thanks for all the suggestions.

I will work on these issues ASAP.

Il 22 settembre 2023 01:13:19 CEST, Dominic Clark ***@***.***> ha scritto:

File: src/core/DataFile.cpp (hunk 4 above)
Comment (mirk0dex) [reply src/core/DataFile.cpp:1862]: This requires me to include <iterator>. Should I?

File: include/DataFile.h [hunk 8]
```diff
@@ -127,8 +127,8 @@ class LMMS_EXPORT DataFile : public QDomDocument
 	void upgrade_mixerRename();
 	void upgrade_bbTcoRename();
 	void upgrade_sampleAndHold();
-  void upgrade_midiCCIndexing();
 	void upgrade_loopsRename();
+  	void upgrade_midiCCIndexing();
```
Comment (zonkmachine) [inline include/DataFile.h:131]: I believe I actually did tell you earlier to move it to after `upgrade_sampleAndHold()`. I should have been more clear. I meant to move it to the end of the list but now the code has changed so the end of the list is after `upgrade_midiCCIndexing`. I believe this to be the correct place of new upgrade routines but I'm not absolutely sure. I will bump the developers on discord about this. I do think a3cad7c should be reverted however, apart from the fix of spaces to tabs in front of `void upgrade_midiCCIndexing();`.

**Review by zonkmachine — CHANGES_REQUESTED** [review]
Hopefully the last review.

File: include/DataFile.h (hunk 8 above)
Comment (zonkmachine) [reply include/DataFile.h:131]: Update. I got a confirm that the new routine should be placed last in line.

File: src/core/DataFile.cpp [hunk 9]
```diff
@@ -1804,7 +1805,79 @@ void DataFile::upgrade_sampleAndHold()
...
+			{
+				auto item = elements.item(i).toElement();
+
+				if (item.isNull()) { continue; }
+				if (item.hasAttribute(srcAttr))
+				{
+					for (const auto& cur : loopBPMs)
+					{
+						auto x = (QString) cur.first,  // loop name
+							 y = (QString) cur.second, // BPM
```
[the remark this answers is not in the document]
Comment (mirk0dex) [reply src/core/DataFile.cpp:1867]: Yes, my bad

**Comment by zonkmachine** [comment]
> > No. You should make sure that things build before committing it to a project or be sure to mention that you haven't tested it yet.
> 
> I did say it in a previous message. Apologies for the inconvenience. Anyways, tomorrow I will have my Arch system with me, so I will be able to compile everything with ease.

For the record, yes you did. I came through as a bit more salty than intended. Sorry!

Thanks for contributing to lmms and for sticking with us through the review process.

**Comment by mirk0dex** [comment]
No worries, I understand!

It was fun to contribute to the project, thanks to you all for this amazing piece of software.

**Pull request merged**